DOC-13982: Eventing Fine-grained RBAC (#654)
couchbaselabs/docs-devex: modules/eventing/pages/eventing-rbac.adoc

{description}

[#description]
== Role-Based Access Control (RBAC)

Couchbase provides _Role-Based Access Control_ (RBAC), in which access privileges are assigned to fixed roles, which are in turn assigned to users, (each of which may be an administrator or an application) either _directly_ or _indirectly_, through _user-groups_.

A Couchbase _role_ permits one or more _resources_ to be accessed according to defined _privileges_.
Roles can be assigned to individual users, and to groups, by means of either the UI or the REST API. 

Couchbase Enterprise edition provides RBAC with multiple roles for finer access control. 

In Capella, RBAC is implemented at the organization, project, and cluster levels.

For a complete list of roles, see xref:learn:security/roles.adoc[Roles].

NOTE: Most roles can be assigned only on the _Enterprise Edition_ of Couchbase Server: on the _Community Edition_ of Couchbase Server, only the `bucket_full_access`, `admin`, and `ro_admin` roles can be assigned.

For more information, see xref:learn:security/authorization-overview.adoc[Authorization], and xref:manage:manage-security/manage-users-and-roles.adoc[Manage Users, Groups, and Roles].

> To learn more about Eventing RBAC, Function Scope, privileged and non-privileged users, and eventing RBAC roles see xref:eventing:eventing-rbac.adoc[Eventing Role-Based Access Control].

Reviewer: Delete this entire **Role-Based Access Control (RBAC)** section. It's all very Server-specific. The Capella documentation already contains full information about how access works, in a way which is tailored for Capella. (As a minor point, it's worth noting that the Capella documentation doesn't usually even refer to RBAC; it mostly uses the term access credentials.)

////
Writer's Note

I want to provide a link to the appropriate Server documentation. Will have to check how to do that and replace the above link to the Eventing RBAC Server doc.

Should we also include all the relevant RBAC information like Roles, privileges etc that are discussed in the Server documentation, here as well? or is it sufficient to just link to the Server documentation?

> ////

Reviewer: Delete the writer's note. There's no need to link to the Server section from here, it would just confuse the users.

> == Access Credentials

Reviewer: Delete the whole **Access Credentials** section, including the subsections on **Access Credential Types**, **Basic Access Credentials**, and **Advanced Access Credentials**. It's all explained already in [Manage Users and Cluster Access](https://docs.couchbase.com/cloud/organizations/organization-projects-overview.html).

== Access Credential Types

The available Eventing access credential types are basic access credentials and advanced access credentials.

=== Basic Access Credentials
Allows you to assign read, write, or read/write access at the bucket, scope, and collection level.

For example, you can grant access to all buckets and scopes in a cluster, assign different access levels to individual buckets, or grant access to just a single collection. 

This allows you to mix and match access levels to different buckets, scopes, and collections in a cluster to match your application and security requirements. 
For more information, see the xref:clusters:manage-database-users.adoc[Cluster Access].

=== Advanced Access Credentials
Provides customized combinations of fine-grained privileges and access roles with precise control over individual operations. 

For example, allows you to assign access to the Owner, Project Creator, or Organization Member role, as described in xref:organizations:organization-user-roles.adoc[], or to the Project Owner or Data Writer role as described in xref:projects:project-roles.adoc[Project Roles]. 

> === Matrix for Eventing UI Access 

Reviewer: Change this to a level 2 heading and a more descriptive title

```suggestion
== Access Eventing with the Capella UI 
```

> If the client is using Basic access credentials, then the user must have the "Eventing Full Admin" role, that provides full access to all Eventing functions and resources.

Reviewer: This section is not about basic access credentials, it's about project roles.

```suggestion
To be able to use the Eventing service through the Capella UI, you must have one of the following project roles: Data Reader, Data Writer, or Project Owner.
For more information, see xref:projects:project-roles.adoc[].
```

> The following table summarizes the advanced access credentials and the mapping to be set for Couchbase Eventing Capella users for managing the Eventing Functions. 

Reviewer: This table is not about advanced access credentials, it's about project roles.

```suggestion
The following table summarizes the actions that you can perform with the Eventing Service using each of these project roles.
```

> [#table-eventing-uiaccess-privileges, cols="1,2,3,4",options="header"]

Reviewer: The `cols` attribute is for specifying the column formatting. You've specified that the first column should be 10% of the table width, the second column should be 20% of the table width, the 3rd column should be 30% of the table width, and the fourth column should be 40% of the table width. This is clearly not what you want. I would suggest that you want columns 2, 3, and 4 to have equal width and column 1 to be twice as wide.

You want the first column to be bold (strong) so specify that here.

```suggestion
[#table-eventing-uiaccess-privileges, cols="2s,1,1,1", options="header"]
```

Author: Thank you for letting me know about this. Noted it down.

|===
| Action 
| Data Reader 
| Data Writer 
| Project Owner 

a|[#create-import-edit]

> `*Create / Import / Edit*` 

Reviewer: These are actions, not code names or variable names. Don't use fixed width text here.

```suggestion
Create / Import / Edit Function
```

a| No 
a| Yes 
a| Yes 

a|[#deploy-undeploy]

> `*Deploy / Undeploy*` 

Reviewer: Not code.

```suggestion
Deploy / Undeploy 
```

a| No 
a| Yes 
a| Yes 

a|[#view-logs-js-settings]

> `*View Logs / JS / Settings*` 

Reviewer: Not code. Also spell out JavaScript in full.

```suggestion
View Logs / JavaScript / Settings 
```

a| Yes 
a| Yes 
a| Yes 

a|[#delete-function]
`*Delete Function*` 
a| No 
a| *Check* 

> a| Yes 

Reviewer: Not code. Also why is "check" left in here? My guess is this should be Yes.

```suggestion
Delete Function 
a| No 
a| Yes 
a| Yes 
```

a|[#export-function]

> `*Export Function*`

Reviewer: Not code

```suggestion
Export Function
```

a| Yes
a| Yes
a| Yes 

|=== 

NOTE: The Cluster Viewer and Cluster Manager roles do not grant Eventing privileges. 

> The "Eventing / Manage Scope Function" role can be assigned to a user multiple times, each time with a different bucket.scope or bucket.

Reviewer: Delete the second sentence in this note, it's got nothing to do with project roles.

```suggestion
NOTE: The Cluster Viewer and Cluster Manager roles do not grant Eventing privileges.
```

> === Programmatic Access via SDK / API

Reviewer: Change this to a level 2 heading and a more descriptive title

```suggestion
== Access Eventing with Cluster Access Credentials
```

To prevent unauthorized access to your Eventing functions, you must use appropriate cluster access credentials with the necessary access levels or roles for your client applications. 

> Data DCP Reader and Eventing / Manage Functions access are required to read from or write to the resources.

Reviewer: Update to give a better description. Also remove the names of the advanced access credential privileges; we'll incorporate them into the table for easier understanding.

```suggestion
To access the Evening Service programmatically via an SDK or API, your client must have the appropriate cluster access credentials, with access to the buckets, scopes, and collections that your Eventing functions use.
For more information, see xref:clusters:cluster-rbac.adoc[].
```

> The following table summarizes the keyspace access permissions to be set for Couchbase Eventing users for managing the Eventing functions via SDK or API. 

Reviewer: Update description for clarity

```suggestion
The following table summarizes the basic access levels or advanced access privileges that your cluster access credentials must have, for each of the target keyspaces that your Eventing functions read from or write to. 
```

> [#table-eventing-keyspace-required-privileges, cols="1,2",options="header"]

Reviewer: Let's add another column for advanced cluster access credentials. Again, you want the first column to be wider than the other two.

```suggestion
[#table-eventing-keyspace-required-privileges, cols="2s,1,1", options="header"]
```

|===
| Target Keyspace 

> | Required Role / Privilege 

Reviewer: Adding advanced access credentials

```suggestion
| Basic Access Level
| Advanced Access Privilege 
```

a|[#source-mutation]
`*Source / Mutation*`

> a| Read or Read / Write 

Reviewer: The target keyspaces are not code names. But the basic access levels are. Add advanced access credentials.

```suggestion
Source / Mutation
a| `Read` or `Read / Write`
a| `Data Read` and `Eventing Manage` 
```

a|[#metadata-storage]
`*Metadata / Storage*` 

> a| Read or Write 

Reviewer: Target keyspaces are not code names. But basic access levels are. Correct the basic access level so it's what SME actually said. Add advanced access credentials.

```suggestion
Metadata / Storage
a| `Read / Write`
a| `Data Read` and `Eventing Manage`
```

a|[#bindings]
`*Bindings*` 

> a| Read or Read / Write 

Reviewer: Target keyspaces are not code names. But basic access levels are. Add advanced access credentials.

```suggestion
Bindings
a| `Read` or `Read / Write`
a| `Data Read` and `Eventing Manage`
```

|===

== See Also

* xref:clusters:cluster-rbac.adoc[]
* xref:clusters:manage-database-users.adoc[]
* xref:organizations:organizations.adoc[]
* xref:organizations:organization-user-roles.adoc[]
* xref:projects:projects.adoc[]
* xref:projects:project-roles.adoc[]
* xref:clusters:manage-database-users.adoc[]
* xref:clusters:allow-ip-address.adoc[]

////
Writer: Not sure if the above references are required. 
Kindly review and let me know if we need to have them!

> ////

Reviewer: I think you only need to link to a few higher-level headings.

```suggestion
* xref:organizations:organizations.adoc[]
* xref:projects:projects.adoc[]
* xref:clusters:cluster-rbac.adoc[]
```
